### PR TITLE
Add Exception handling and test for divide by zero exception.

### PR DESCRIPTION
**Pull request by jplevyak**
Add Exception handling and test for divide by zero exception.

### CODE REVIEW

File: source/extensions/common/wasm/wavm/wavm.cc [hunk 1]
```diff
@@ -73,13 +73,27 @@ struct SaveRestoreContext {
 #define CALL_WITH_CONTEXT(_x, _context)                                                            \
   do {                                                                                             \
     SaveRestoreContext _saved_context(static_cast<Context*>(_context));                            \
-    _x;                                                                                            \
+    WAVM::Runtime::catchRuntimeExceptions([&] {                                                     \
+      _x;                                                                                          \
+    }, [&](WAVM::Runtime::Exception* exception) {                                              \
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/wavm/wavm.cc]: Nit: `\` seem all over the place, could you align them?

File: source/extensions/common/wasm/wavm/wavm.cc [hunk 2]
```diff
@@ -73,13 +73,27 @@ struct SaveRestoreContext {
...
+      auto description = describeException(exception);                                            \
+      delete exception; \
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/wavm/wavm.cc]: While this might be equivalent at this time, I think you should call `destroyException(exception)` here.

File: test/extensions/wasm/wasm_test.cc [hunk 3]
```diff
@@ -73,6 +73,52 @@ TEST(WasmTest, BadSignature) {
...
+  const auto code = TestEnvironment::readFileToStringForTest(
+      TestEnvironment::substitute("{{ test_rundir }}/test/extensions/wasm/test_data/segv.wasm"));
+  EXPECT_FALSE(code.empty());
+  auto context = std::make_unique<TestContext>(wasm.get());
+  EXPECT_CALL(*context, scriptLog(spdlog::level::err, Eq("before div by zero")));
+  EXPECT_TRUE(wasm->initialize(code, "<test>", false));
+  wasm->setGeneralContext(std::move(context));
+  wasm->wasmVm()->start(wasm->generalContext());
+#if 0
+  // Not yet supported in gtes.
```
Comment (PiotrSikora) [inline test/extensions/wasm/wasm_test.cc]: gtes?

File: test/extensions/wasm/wasm_test.cc [hunk 4]
```diff
@@ -73,6 +73,52 @@ TEST(WasmTest, BadSignature) {
...
+  EXPECT_THROW_WITH_MESSAGE_THAT(wasm->generalContext()->onLog(), Extensions::Common::Wasm::WasmException,
+                                 testing::StartsWith("wavm.integerDivideByZeroOrOverflow"));
+#endif
+  EXPECT_THROW(wasm->generalContext()->onLog(), Extensions::Common::Wasm::WasmException);
+      // testing::Property(&std::exception::what(), testing::StartsWith("wavm.integerDivideByZeroOrOverflow")));
```
Comment (PiotrSikora) [inline test/extensions/wasm/wasm_test.cc]: Nit: debug leftover?

File: test/extensions/wasm/wasm_test.cc [hunk 5]
```diff
@@ -73,6 +73,52 @@ TEST(WasmTest, BadSignature) {
...
+      TestEnvironment::substitute("{{ test_rundir }}/test/extensions/wasm/test_data/segv.wasm"));
+  EXPECT_FALSE(code.empty());
+  auto context = std::make_unique<TestContext>(wasm.get());
+  EXPECT_CALL(*context, scriptLog(spdlog::level::err, Eq("before div by zero")));
+  EXPECT_TRUE(wasm->initialize(code, "<test>", false));
+  wasm->setGeneralContext(std::move(context));
+  wasm->wasmVm()->start(wasm->generalContext());
+#if 0
+  // Not yet supported in gtes.
+  EXPECT_THROW_WITH_MESSAGE_THAT(wasm->generalContext()->onLog(), Extensions::Common::Wasm::WasmException,
```
Comment (PiotrSikora) [inline test/extensions/wasm/wasm_test.cc]: Nit: `EXPECT_THROW_WITH_MESSAGE` not `EXPECT_THROW_WITH_MESSAGE_THAT`.

File: ci/build_container/build_recipes/wavm.sh [hunk 6]
```diff
@@ -31,8 +31,8 @@ cp -pR include/ "$THIRDPARTY_BUILD"/lib/llvm-6.0/include/
 
 # WAVM.
 
-COMMIT=69d647ecf6d355921b79a58aa6b1d674a76b4c2b # 2019-01-24
-SHA256=b0789161eb96cc0695cda6e0ea0b2db0bd1919795cebf5d139d66a173234d301
+COMMIT=d3607084b98017243cefe699064d4d5cf3f50fd9  # Thu Feb 14 13:12:32 2019 -0500
```
Comment (PiotrSikora) [inline ci/build_container/build_recipes/wavm.sh]: Same comment as in the other PR, why not latest version (i.e. `275e15fff5928d5cac3b8c5410c5a3b3fa7168ac`)?

File: source/extensions/common/wasm/wasm.cc [hunk 7]
```diff
@@ -4,6 +4,7 @@
 
 #include <memory>
 #include <string>
+#include <vector>
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/wasm.cc]: Why is this needed?

File: source/extensions/common/wasm/wavm/wavm.cc (hunk 1 above)
Comment (jplevyak) [reply source/extensions/common/wasm/wavm/wavm.cc]: Fixed all formatting.  Done.

File: source/extensions/common/wasm/wavm/wavm.cc (hunk 2 above)
Comment (jplevyak) [reply source/extensions/common/wasm/wavm/wavm.cc]: done.

File: test/extensions/wasm/wasm_test.cc (hunk 3 above)
Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: gtest done.

File: test/extensions/wasm/wasm_test.cc (hunk 5 above)
Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: The problem here is that `EXPECT_THROW_WITH_MESSAGE` doesn't support matchers, and I need to do a string prefix.  EXPECT_THROW_WITH_MESSAGE_THAT is in a proposed PR for gtest/gmock.

I compromised and I just check that we were able to translate the WAVM Exception into an envoy WasmException.

Comment (PiotrSikora) [reply test/extensions/wasm/wasm_test.cc]: Wouldn't `EXPECT_THROW_WITH_REGEX()` work? i.e.
```
EXPECT_THROW_WITH_REGEX(wasm->generalContext()->onLog(), Extensions::Common::Wasm::WasmException,
                        "wavm.integerDivideByZeroOrOverflow.*");
```

File: ci/build_container/build_recipes/wavm.sh (hunk 6 above)
Comment (jplevyak) [reply ci/build_container/build_recipes/wavm.sh]: Done.

File: source/extensions/common/wasm/wasm.cc (hunk 7 above)
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: Removed.  Left over from some merge/patch I presume.

File: test/extensions/wasm/wasm_test.cc (hunk 5 above)
Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: The reason I need the string prefix is that the stack trace is in the remained of the message.

Comment (PiotrSikora) [reply test/extensions/wasm/wasm_test.cc]: Yeah, hence the `".*"` (regex) to catch reminder of the output.

Comment (PiotrSikora) [reply test/extensions/wasm/wasm_test.cc]: FWIW, I've added the snippet from my previous comment to the test and it passes.

Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: Didn't see you message.  Yes that works, Done.
